### ISSUE TITLE
how can I generate images in figure 3?

### ISSUE THREAD

**Issue by seyeeet** [SsnL/align_uniform#4]
Hi
Thanks for the cool work!
Can you please guide me how to generate/produce the experiments and visualization for Fig 3?
Thanks

**Comment by ssnl** [comment]
Hi! Thanks for your interest in the paper. The code uses some internal utility library so I can't share it. But here is the rough idea:

+ Vectors are on the unit circle S^1 (2-dimensional)

+ validation set augmentations are as follows:

  Positive pairs (used to measure alignment) are sampled with 
```
  torchvision.transforms.RandomResizedCrop(64, scale=(0.3, 1.0), ratio=(0.7, 1.4),interpolation=Image.BICUBIC), 
  torchvision.transforms.RandomHorizontalFlip(), 
  torchvision.transforms.ColorJitter(0.4, 0.4, 0.4, 0.4), 
  torchvision.transforms.RandomGrayscale(p=0.2),  
  torchvision.transforms.Normalize((0.4914, 0.4822, 0.4465), (0.247, 0.243, 0.261)),
```

 uniformity are measured with   
```
  torchvision.transforms.Resize(64),
  torchvision.transforms.Normalize((0.4914, 0.4822, 0.4465), (0.247, 0.243, 0.261)), 
```
+ use vMF KDE with kappa=25

Let me know if you have any questions!

**Comment by seyeeet** [comment]
Thank you! I will give it a try and let you know if I have any question.

**Comment by lichen14** [comment]
@SsnL Hi, thanks for sharing.
I am wondering how to get the unit circle S^1 vectors, can u give me some ideas?  so that I can plot the 2d vectors in a circle.
 : )

**Comment by ssnl** [comment]
For the plot encoders are modified to directly output normalized 2d
vectors.

**Comment by lichen14** [comment]
> For the plot encoders are modified to directly output normalized 2d vectors.
> […](#)

Thanks for the reply. 
Do u mean that I need convert the normalized 2d vectors to polar coordinates, in order to get the circle feature distribution? or KDE can achieve this? ( I am not familiar with KDE :D )

**Comment by ssnl** [comment]
It is not really about KDEs. The encoder outputs are (normalized) 2D vectors which are directly on S^1. The plot is for intuitive understanding of the ideas in the paper, not for general encoders whose output dimensionality is often 128, 256, or more.

**Comment by YunYunY** [comment]
Hi thank you for sharing the code and detail. When you say modifying the output to 2D, do you mean changing the last layer of encoder output from 256 to 2 and training the representation network from scratch?

**Comment by ssnl** [comment]
@YunYunY Yes.

**Comment by YunYunY** [comment]
Thanks, I have a silly following question. If as Figure 3 showed 2D representation is well separable, why do we usually use 256 embeddings for self-supervised pretraining?

**Comment by ssnl** [comment]
@YunYunY From Fig 3, I would say that the feature distributions for difference classes on S^1 still overlap a lot, and are not as cleanly separable. You can also see that from the low linear classification accuracy 28%, compared to the much higher numbers one can get on CIFAR-10 w/ larger latent spaces.

**Comment by YunYunY** [comment]
Ah yes, thanks

**Comment by liesun1994** [comment]
Hi, @SsnL. As you mentioned above, we need to use vMF KDE. Could u please share the code to calculate the vMT KDE with the given normalized vector? It could really helpful.

**Comment by ssnl** [comment]
> Hi, @SsnL. As you mentioned above, we need to use vMF KDE. Could u please share the code to calculate the vMT KDE with the given normalized vector? It could really helpful.

I don't have the code now but just googling "con mises fisher kde Python" should give you plenty of ideas.
